optics-core/src/Optics/Indexed.hs

```diff
@@ -30,67 +34,153 @@ import Optics.Internal.Optic
 import Optics.Internal.Optic.TypeLevel
 import Optics.Internal.Profunctor
 
-import Optics.IxTraversal
+import Optics.Fold
 import Optics.IxFold
 import Optics.IxSetter
+import Optics.IxTraversal
+import Optics.Setter
+import Optics.Traversal
+
+-- | Compose two indexed optics. Their indices are composed as a pair.
+--
+-- >>> itoListOf (ifolded <%> ifolded) ["foo", "bar"]
+-- [((0,0),'f'),((0,1),'o'),((0,2),'o'),((1,0),'b'),((1,1),'a'),((1,2),'r')]
+--
+infixr 9 <%>
+(<%>)
+  :: (m ~ Join k l, Is k m, Is l m, IxOptic m s t a b, '[i, j] ~ Append is js)
+  => Optic k is        s t u v
+  -> Optic l js        u v a b
+  -> Optic m '[(i, j)] s t a b
```

Comment:
Eh, there is a bunch of singleton lists in this module. Perhaps we should always use WithIx instead of singleton list?

Reply:
Yes, we should use `WithIx` whenever possible.

Reply:
Also, this works when `is = '[i, j]` and `js = '[]`.

And given fixities, will `folded <%> ifolded % ifolded` work? We could either document this quirk, or use `CheckIndices`.

Reply:
Hmm, right, I think that type signature is too general.

```haskell
(<%>)
  :: (m ~ Join k l, Is k m, Is l m, IxOptic m s t a b)
  => Optic k '[i]      s t u v
  -> Optic l '[j]      u v a b
  -> Optic m '[(i, j)] s t a b
```
is probably better.

with current signature both
```haskell
λ> :t folded % ifolded <%> ifolded
folded % ifolded <%> ifolded
  :: (FoldableWithIndex i1 f1, FoldableWithIndex i2 f2,
      Foldable f3) =>
     Optic A_Fold '[(i1, i2)] (f3 (f1 (f2 b))) (f3 (f1 (f2 b))) b b
λ> :t folded <%> ifolded % ifolded
folded <%> ifolded % ifolded
  :: (FoldableWithIndex i1 f1, FoldableWithIndex i2 f2,
      Foldable f3) =>
     Optic A_Fold '[(i1, i2)] (f3 (f1 (f2 b))) (f3 (f1 (f2 b))) b b
```
work but with restricted one only the former type checks:
```haskell
λ> :t folded % ifolded <%> ifolded
folded % ifolded <%> ifolded
  :: (FoldableWithIndex i f1, FoldableWithIndex j f2, Foldable f3) =>
     Optic A_Fold '[(i, j)] (f3 (f1 (f2 b))) (f3 (f1 (f2 b))) b b
λ> :t folded <%> ifolded % ifolded

<interactive>:1:1: error:
    • Couldn't match type ‘'[]’ with ‘'[i]’
      Expected type: Optic
                       A_Fold '[i] (f (f1 (f2 b))) (f (f1 (f2 b))) (f1 (f2 b)) (f1 (f2 b))
        Actual type: Fold (f (f1 (f2 b))) (f1 (f2 b))
    • In the first argument of ‘(<%>)’, namely ‘folded’
      In the expression: folded <%> ifolded % ifolded

<interactive>:1:12: error:
    • Couldn't match type ‘'[]’ with ‘'[i0]’ arising from a use of ‘%’
    • In the second argument of ‘(<%>)’, namely ‘ifolded % ifolded’
      In the expression: folded <%> ifolded % ifolded
```

I think this is fine, the latter expression looks weird.

Reply:
I'm also not set on fixity. `infixr 9` is what lens use, but it's somewhat forced too as lens operators need to have the same fixity as `(.)`, we have more choice. Perhaps `infixl 9` would be better.

Reply:
I restricted the signature.

Reply:
I'd still rather use `CheckIndices is i` to get better error messages!

Reply:
And about fixities: I think `infixl` would be more natural indeed. @adamgundry @kosmikus opinions?

Reply:
I wasn't sold on using CheckIndices there because the error message wouldn't point you to where the trouble is, but I modified the definition a bit and now it works much better:

```haskell
λ> :t folded <%> ifolded % ifolded

<interactive>:1:1: error:
    • Indexed optic is expected
        (in the first argument of ‘(<%>)’)
    • In the expression: folded <%> ifolded % ifolded

<interactive>:1:1: error:
    • Use (<%>) or icompose to combine indices of type ‘j’ and ‘i0’
        (in the second argument of ‘(<%>)’)
    • In the expression: folded <%> ifolded % ifolded

λ> :t itraverseOf_ (folded <%> ifolded % ifolded)

<interactive>:1:15: error:
    • Indexed optic is expected
        (in the first argument of ‘(<%>)’)
    • In the first argument of ‘itraverseOf_’, namely
        ‘(folded <%> ifolded % ifolded)’
      In the expression: itraverseOf_ (folded <%> ifolded % ifolded)

<interactive>:1:15: error:
    • Use (<%>) or icompose to combine indices of type ‘j’ and ‘i0’
        (in the second argument of ‘(<%>)’)
    • In the first argument of ‘itraverseOf_’, namely
        ‘(folded <%> ifolded % ifolded)’
      In the expression: itraverseOf_ (folded <%> ifolded % ifolded)
```

As for the fixity, let's leave infixr for now and merge the PR, we can discuss in #20.